File: coconut/env/oracle.go
package env

import (
	"errors"
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
)

type Oracle struct {
	home           string
	sid            string
	user           string
	password       string
	directory_path string
}

func NewOracle(home string, sid string, user string, password string, directory_path string) (*Oracle, error) {
	{
		fi, err := os.Stat(filepath.Join(home, "bin", "impdp"))
		if err != nil {
			return nil, err
		}
		if fi.IsDir() {
			return nil, errors.New("bad oracle home")
		}
	}
	if sid == "" {
		return nil, errors.New("empty SID")
	}
	if user == "" {
		return nil, errors.New("empty user")
	}

	return &Oracle{home: home, sid: sid, directory_path: directory_path, user: user, password: password}, nil
}

func (p *Oracle) Dump(target string) []*exec.Cmd {
	slog.Info("backup oracle", slog.String("sid", p.sid), slog.String("user", p.user))
	slog.Warn(fmt.Sprintf(`
Please make sure:
create or replace directory datapump_dir as '%s';
grant read,write on directory datapump_dir to %s;
grant datapump_exp_full_database to %s;
grant datapump_imp_full_database to %s;
`, p.directory_path, p.user, p.user, p.user))

	exp := exec.Command(
		filepath.Join(p.home, "bin", "expdp"),
		fmt.Sprintf("%s/%s", p.user, p.password),
		"full=Y",
		fmt.Sprintf("dumpfile=%s.dmp", p.user),
		fmt.Sprintf("logfile=%s-exp.log", p.user),
	)
	exp.Env = append(os.Environ(),
		fmt.Sprintf("ORACLE_HOME=%s", p.home),
		fmt.Sprintf("ORACLE_SID=%s", p.sid),
		"NLS_LANG=AMERICAN_AMERICA.AL32UTF8",
		"LANG=en_US.utf8",
	)
	return []*exec.Cmd{
		exp,
		exec.Command("mv", filepath.Join(p.directory_path, fmt.Sprintf("%s.dmp", p.user)), filepath.Join(p.directory_path, fmt.Sprintf("%s-exp.log", p.user)), target),
	}
}
func (p *Oracle) Restore() []*exec.Cmd {
	return []*exec.Cmd{
		exec.Command("mv", fmt.Sprintf("%s.dmp", p.user), p.directory_path),
		exec.Command(
			filepath.Join(p.home, "bin", "impdp"),
			fmt.Sprintf("%s/CHANGE-ME", p.user),
			fmt.Sprintf("dumpfile=%s.dmp", p.user),
			fmt.Sprintf("logfile=%s.exp.log", p.user),
			fmt.Sprintf("schemas=%s", p.user),
			fmt.Sprintf("REMAP_SCHEMA=%s:NEW-USER", p.user),
		),
	}
}

File: aloe/src/oracle.cpp
#include "aloe/oracle.hpp"
#include "palm/crypto.hpp"
#include "palm/utils.hpp"

void aloe::Oracle::dump() {
  const std::string tmp = std::format("{}-{}", this->_sid, palm::timestamp());
  const std::string tar = std::format("{}.tar", tmp);
  spdlog::info("dump oracle://{}@:127.0.0.1:1521/{} to {}/{}", this->_user,
               this->_sid, this->_directory_path, tmp);
  const std::string log = std::format("{}.exp.log", tmp);
  const std::string dmp = std::format("{}.dmp", tmp);

  {
    const auto& [status, out, err] = palm::shell(
        "/usr/bin/expdp",
        {std::format("{}/{}", this->_user, this->_password), "full=Y",
         std::format("dumpfile=", dmp), std::format("logfile=", log)});
    if (status != EXIT_SUCCESS) {
      spdlog::error("{} {}", status, err);
      return;
    }
    spdlog::debug("{}", out);
  }

  {
    const auto& [status, out, err] =
        palm::tar(tar, this->_directory_path, {log, dmp});
    if (status != EXIT_SUCCESS) {
      spdlog::error("{} {}", status, err);
      return;
    }
    spdlog::debug("{}", out);
  }
  {
    const auto& [status, out, err] = palm::xz(tar);
    if (status != EXIT_SUCCESS) {
      spdlog::error("{} {}", status, err);
      return;
    }
    spdlog::debug("{}", out);
  }
  spdlog::info("done({}.xz).", tmp);
}

void aloe::Oracle::restore(const std::string& name, const std::string& user) {
  spdlog::info("restore {} to oracle://{}@:127.0.0.1:1521/{}", name,
               this->_user, this->_sid);
  const std::string log = std::format("{}.imp.log", name);
  const std::string dmp = std::format("{}.dmp", name);
  {
    const auto& [status, out, err] = palm::shell(
        "/usr/bin/xz", {"--decompress", std::format("{}.tar.xz", name)});
    if (status != EXIT_SUCCESS) {
      spdlog::error("{} {}", status, err);
      return;
    }
    spdlog::debug("{}", out);
    std::filesystem::rename(dmp,
                            std::filesystem::path(this->_directory_path) / dmp);
  }
  {
    const auto& [status, out, err] = palm::shell(
        "/usr/bin/impdp",
        {std::format("{}/{}", this->_user, this->_password), "full=Y",
         std::format("dumpfile={}", dmp), std::format("logfile={}", log),
         std::format("schemas={}", this->_user),
         std::format("remap_schema={}:{}", this->_user, user)});
    if (status != EXIT_SUCCESS) {
      spdlog::error("{} {}", status, err);
      return;
    }
    spdlog::debug("{}", out);
  }
  {
    std::filesystem::path it =
        std::filesystem::path(this->_directory_path) / dmp;
    spdlog::debug("remove {}", it.string());
    std::filesystem::remove(it);
  }
  {
    const auto& [status, out, err] =
        palm::xz(std::filesystem::path(this->_directory_path) / log);
    if (status != EXIT_SUCCESS) {
      spdlog::error("{} {}", status, err);
      return;
    }
    spdlog::debug("{}", out);
  }
  spdlog::info("done.");
}
